Error exporting a project in XLSX

**Describe the bug**
In OR 3.3 beta, I got a java error message when I tried to export a project into XLSX

**To Reproduce**

**Current Results**

Here is the error message

HTTP ERROR 500

Problem accessing /command/core/export-rows/get_q_any_contains.xlsx. Reason:

    Address of hyperlink must be a valid URI

Caused by:

```
java.lang.IllegalArgumentException: Address of hyperlink must be a valid URI
	at org.apache.poi.xssf.usermodel.XSSFHyperlink.validate(XSSFHyperlink.java:264)
	at org.apache.poi.xssf.usermodel.XSSFHyperlink.setAddress(XSSFHyperlink.java:245)
	at com.google.refine.exporters.XlsExporter$1.addRow(XlsExporter.java:132)
	at com.google.refine.exporters.CustomizableTabularExporterUtilities$1.visit(CustomizableTabularExporterUtilities.java:168)
	at com.google.refine.browsing.util.RowVisitorAsRecordVisitor.visit(RowVisitorAsRecordVisitor.java:61)
	at com.google.refine.browsing.util.ConjunctiveFilteredRecords.accept(ConjunctiveFilteredRecords.java:64)
	at com.google.refine.browsing.util.FilteredRecordsAsFilteredRows.accept(FilteredRecordsAsFilteredRows.java:50)
	at com.google.refine.exporters.CustomizableTabularExporterUtilities.exportRows(CustomizableTabularExporterUtilities.java:182)
	at com.google.refine.exporters.XlsExporter.export(XlsExporter.java:140)
	at com.google.refine.commands.project.ExportRowsCommand.doPost(ExportRowsCommand.java:119)
	at com.google.refine.RefineServlet.service(RefineServlet.java:189)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at org.mortbay.servlet.UserAgentFilter.doFilter(UserAgentFilter.java:81)
	at org.mortbay.servlet.GzipFilter.doFilter(GzipFilter.java:132)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:938)
	at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:755)
	at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:218)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at org.mortbay.jetty.bio.SocketConnector$Connection.run(SocketConnector.java:228)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.URISyntaxException: Illegal character in query at index 81: http://catalogue.unice.fr/primo-explore/fulldisplay?docid=sc_aleph_uns01000411732 any,contains,comptabilité générale zoom
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3111)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at org.apache.poi.xssf.usermodel.XSSFHyperlink.validate(XSSFHyperlink.java:262)
	... 32 more

Caused by:

java.net.URISyntaxException: Illegal character in query at index 81: http://catalogue.unice.fr/primo-explore/fulldisplay?docid=sc_aleph_uns01000411732 any,contains,comptabilité générale zoom
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3111)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at org.apache.poi.xssf.usermodel.XSSFHyperlink.validate(XSSFHyperlink.java:262)
	at org.apache.poi.xssf.usermodel.XSSFHyperlink.setAddress(XSSFHyperlink.java:245)
	at com.google.refine.exporters.XlsExporter$1.addRow(XlsExporter.java:132)
	at com.google.refine.exporters.CustomizableTabularExporterUtilities$1.visit(CustomizableTabularExporterUtilities.java:168)
	at com.google.refine.browsing.util.RowVisitorAsRecordVisitor.visit(RowVisitorAsRecordVisitor.java:61)
	at com.google.refine.browsing.util.ConjunctiveFilteredRecords.accept(ConjunctiveFilteredRecords.java:64)
	at com.google.refine.browsing.util.FilteredRecordsAsFilteredRows.accept(FilteredRecordsAsFilteredRows.java:50)
	at com.google.refine.exporters.CustomizableTabularExporterUtilities.exportRows(CustomizableTabularExporterUtilities.java:182)
	at com.google.refine.exporters.XlsExporter.export(XlsExporter.java:140)
	at com.google.refine.commands.project.ExportRowsCommand.doPost(ExportRowsCommand.java:119)
	at com.google.refine.RefineServlet.service(RefineServlet.java:189)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at org.mortbay.servlet.UserAgentFilter.doFilter(UserAgentFilter.java:81)
	at org.mortbay.servlet.GzipFilter.doFilter(GzipFilter.java:132)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:938)
	at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:755)
	at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:218)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at org.mortbay.jetty.bio.SocketConnector$Connection.run(SocketConnector.java:228)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Powered by Jetty://


Here are the logs in Win10 console :

1:27:40.390 [        ExportRowsCommand] error:Address of hyperlink must be a valid URI (8860ms)
11:27:40.397 [          org.mortbay.log] /command/core/export-rows/get_q_any_contains.xlsx (7ms)
java.lang.IllegalArgumentException: Address of hyperlink must be a valid URI
        at org.apache.poi.xssf.usermodel.XSSFHyperlink.validate(XSSFHyperlink.java:264)
        at org.apache.poi.xssf.usermodel.XSSFHyperlink.setAddress(XSSFHyperlink.java:245)
        at com.google.refine.exporters.XlsExporter$1.addRow(XlsExporter.java:132)
        at com.google.refine.exporters.CustomizableTabularExporterUtilities$1.visit(CustomizableTabularExporterUtilities.java:168)
        at com.google.refine.browsing.util.RowVisitorAsRecordVisitor.visit(RowVisitorAsRecordVisitor.java:61)
        at com.google.refine.browsing.util.ConjunctiveFilteredRecords.accept(ConjunctiveFilteredRecords.java:64)
        at com.google.refine.browsing.util.FilteredRecordsAsFilteredRows.accept(FilteredRecordsAsFilteredRows.java:50)
        at com.google.refine.exporters.CustomizableTabularExporterUtilities.exportRows(CustomizableTabularExporterUtilities.java:182)
        at com.google.refine.exporters.XlsExporter.export(XlsExporter.java:140)
        at com.google.refine.commands.project.ExportRowsCommand.doPost(ExportRowsCommand.java:119)
        at com.google.refine.RefineServlet.service(RefineServlet.java:189)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
        at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
        at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
        at org.mortbay.servlet.UserAgentFilter.doFilter(UserAgentFilter.java:81)
        at org.mortbay.servlet.GzipFilter.doFilter(GzipFilter.java:132)
        at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
        at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
        at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
        at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
        at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
        at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
        at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
        at org.mortbay.jetty.Server.handle(Server.java:326)
        at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
        at org.mortbay.jetty.HttpConnection$RequestHandler.content(HttpConnection.java:938)
        at org.mortbay.jetty.HttpParser.parseNext(HttpParser.java:755)
        at org.mortbay.jetty.HttpParser.parseAvailable(HttpParser.java:218)
        at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
        at org.mortbay.jetty.bio.SocketConnector$Connection.run(SocketConnector.java:228)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.URISyntaxException: Illegal character in query at index 81: http://catalogue.unice.fr/primo-explore/fulldisplay?docid=sc_aleph_uns01000411732 any,contains,comptabilitÚ gÚnÚrale zoom
        at java.net.URI$Parser.fail(URI.java:2848)
        at java.net.URI$Parser.checkChars(URI.java:3021)
        at java.net.URI$Parser.parseHierarchical(URI.java:3111)
        at java.net.URI$Parser.parse(URI.java:3053)
        at java.net.URI.<init>(URI.java:588)
        at org.apache.poi.xssf.usermodel.XSSFHyperlink.validate(XSSFHyperlink.java:262)
        ... 32 more

```

**Expected behavior**
A clear and concise description of what you expected to happen or to show.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Desktop (please complete the following information):**
 - OS: Win10
 - Browser Version: Firefox
 - JRE or JDK Version: JRE 1.8.0_201

**OpenRefine (please complete the following information):**
 - Version 3.3 beta

**Datasets**
If you are allowed and are OK with making your data public, it would be awesome if you can include or attach the data causing the issue or a URL pointing to where the data is.
If you are concerned about keeping your data private, ping us on our [mailing list](https://groups.google.com/forum/#!forum/openrefine)

**Additional context**
Add any other context about the problem here.

--- a/main/src/com/google/refine/exporters/CustomizableTabularExporterUtilities.java
+++ b/main/src/com/google/refine/exporters/CustomizableTabularExporterUtilities.java
@@ -34,6 +34,8 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 package com.google.refine.exporters;
 
 import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
 import java.text.DateFormat;
 import java.text.SimpleDateFormat;
 import java.time.OffsetDateTime;
@@ -371,10 +373,13 @@ abstract public class CustomizableTabularExporterUtilities {
                         if (value instanceof String) {
                             text = (String) value;
                             
-                            if(text.contains(":")) {
-                                if(urlValidator.isValid(text)) {
-                                    link = text;
-                                }
+                            if(text.contains(":") && urlValidator.isValid(text)) {
+                            	// Extra check for https://github.com/OpenRefine/OpenRefine/issues/2213
+                            	try {
+                            		link = new URI(text).toString();
+                            	} catch(URISyntaxException e) {
+                            		;
+                            	}
                             }
                         } else if (value instanceof OffsetDateTime) {
                             text = ((OffsetDateTime) value).format(DateTimeFormatter.ISO_INSTANT);

--- a/main/src/com/google/refine/exporters/XlsExporter.java
+++ b/main/src/com/google/refine/exporters/XlsExporter.java
@@ -39,7 +39,7 @@ import java.time.OffsetDateTime;
 import java.util.List;
 import java.util.Properties;
 
-import org.apache.poi.common.usermodel.Hyperlink;
+import org.apache.poi.ss.usermodel.Hyperlink;
 import org.apache.poi.common.usermodel.HyperlinkType;
 import org.apache.poi.hssf.usermodel.HSSFWorkbook;
 import org.apache.poi.ss.usermodel.Cell;
@@ -127,9 +127,15 @@ public class XlsExporter implements StreamExporter {
                             }
                             
                             if (cellData.link != null) {
-                                Hyperlink hl = wb.getCreationHelper().createHyperlink(HyperlinkType.URL);
-                                hl.setLabel(cellData.text);
-                                hl.setAddress(cellData.link);
+                            	try {
+                            		Hyperlink hl = wb.getCreationHelper().createHyperlink(HyperlinkType.URL);
+                            		hl.setLabel(cellData.text);
+                            		hl.setAddress(cellData.link);
+                            		c.setHyperlink(hl);
+                            	} catch(IllegalArgumentException e) {
+                            		// If we failed to create the hyperlink and add it to the cell,
+                            		// we just use the string value as fallback
+                            	}
                             }
                         }
                     }

--- a/main/tests/server/src/com/google/refine/exporters/XlsxExporterTests.java
+++ b/main/tests/server/src/com/google/refine/exporters/XlsxExporterTests.java
@@ -43,6 +43,7 @@ import static org.mockito.Mockito.mock;
 
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.Serializable;
 import java.time.OffsetDateTime;
 import java.util.Properties;
 
@@ -155,6 +156,30 @@ public class XlsxExporterTests extends RefineTest {
             Assert.fail();
         }
     }
+    
+    @Test
+    public void exportXlsxStringWithURLs() throws IOException{
+        String url = "GET /primo-library/,http:%2F%2Fcatalogue.unice.fr HTTP/1.1";
+        createDateGrid(2, 2, url);
+
+        try {
+            SUT.export(project, options, engine, stream);
+        } catch (IOException e) {
+            Assert.fail();
+        }
+        
+        ByteArrayInputStream inStream = new ByteArrayInputStream( stream.toByteArray() );
+        try {
+            XSSFWorkbook wb = new XSSFWorkbook(inStream);
+            XSSFSheet ws = wb.getSheetAt(0);
+            XSSFRow row1 = ws.getRow(1);
+            XSSFCell cell0 = row1.getCell(0);
+            Assert.assertTrue(cell0.toString().contains("primo-library"));
+            wb.close();
+        } catch (IOException e) {
+            Assert.fail();
+        }
+    }
 
     //helper methods
 
@@ -180,13 +205,13 @@ public class XlsxExporterTests extends RefineTest {
         }
     }
     
-    private void createDateGrid(int noOfRows, int noOfColumns, OffsetDateTime now){
+    private void createDateGrid(int noOfRows, int noOfColumns, Serializable value){
         CreateColumns(noOfColumns);
 
         for(int i = 0; i < noOfRows; i++){
             Row row = new Row(noOfColumns);
             for(int j = 0; j < noOfColumns; j++){
-                row.cells.add(new Cell(now, null));
+                row.cells.add(new Cell(value, null));
             }
             project.rows.add(row);
         }

